src/ArmTemplates/Common/API/Clients/Abstractions/IApisClient.cs

```diff
@@ -6,15 +6,17 @@
 
 using System.Collections.Generic;
 using System.Threading.Tasks;
-using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Templates.Service;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Templates.Apis;
 using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Extractor.Models;
 
 namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.API.Clients.Abstractions
 {
     public interface IApisClient
     {
-        Task<ApiTemplateResource> GetSingleAsync(string apiManagementName, string resourceGroupName, string apiName);
+        Task<ApiTemplateResource> GetSingleAsync(ExtractorParameters extractorParameters, string apiName);
```

Comment:
`ExtractorParameters` is an immutable container for parameters of extraction. Basically it contains user requested parameters, like `api management instance name`, `api management's resource group` and etc. I am trying to make `ExtractorParameters` be used all over the program, so that you can easily access any user-requested parameter at any moment.

But ApiName is a specific name of API that is desired to be retrieved at the moment. So imagine you are doing some template-generation and eventually you know that your product has `my-super-api-name-1` and `my-crazy-api-name-2` connected to the product. And you want some info for them. Then you want to use `ApisClient` and get information about specific API.

Reply:
If I understand this correctly, you want the ExtractorParameters to be immutable and reuse them across multiple exctractions? If the name/ID is included in the params, then they can't be reused?

Reply:
What about passing the `ExtractorParameters` to the Client constructor instead of passing it to every individual method if it's immutable anyway? That way you create your Clients with some specific settings and you would simply call `GetSingleAsync(string apiName)`.

Reply:
The trouble is that we do not have extractor parameters at the point of registering the services.
I can try to work that out, but it is not proper way for me because user will need to register extractor parameters also in IOC. It is not very good in my opinion - it is not an injectable service, it is a setting. 

I do not know - maybe it is okay to use Microsoft .Extensions.Options.IOptions<T> interface and store extractor parameters there... But still it will be painful to refactor the serviceCollection building stuff only for that purpose

Reply:
It was just an idea :) We don't have to do it if it's hard and not worth it.

Reply:
Okay, I would like to stay with current implementation for now, maybe we can change it in future